lib/pos_snapshot_entries.go

```diff
@@ -0,0 +1,643 @@
+	for snapshotAtEpochNumber, globalParamsEntry := range bav.SnapshotGlobalParamEntries {
+		if globalParamsEntry == nil {
+			return fmt.Errorf(
+				"_flushSnapshotGlobalParamsEntryToDb: found nil entry for EpochNumber %d, this should never happen",
+				snapshotAtEpochNumber,
+			)
+		}
+		if err := DBPutSnapshotGlobalParamsEntryWithTxn(
+			txn, bav.Snapshot, globalParamsEntry, snapshotAtEpochNumber, blockHeight,
+		); err != nil {
+			return errors.Wrapf(
+				err,
+				"_flushSnapshotGlobalParamsEntryToDb: problem setting SnapshotGlobalParamsEntry for EpochNumber %d: ",
+				snapshotAtEpochNumber,
+			)
+		}
+	}
+	return nil
+}
+
+func DBKeyForSnapshotGlobalParamsEntry(snapshotEpochNumber uint64) []byte {
+	key := append([]byte{}, Prefixes.PrefixSnapshotGlobalParamsEntry...)
+	key = append(key, UintToBuf(snapshotEpochNumber)...)
+	return key
+}
+
+func DBGetSnapshotGlobalParamsEntry(handle *badger.DB, snap *Snapshot, snapshotAtEpochNumber uint64) (*GlobalParamsEntry, error) {
+	var ret *GlobalParamsEntry
+	err := handle.View(func(txn *badger.Txn) error {
+		var innerErr error
+		ret, innerErr = DBGetSnapshotGlobalParamsEntryWithTxn(txn, snap, snapshotAtEpochNumber)
+		return innerErr
+	})
+	return ret, err
+}
+
+func DBGetSnapshotGlobalParamsEntryWithTxn(txn *badger.Txn, snap *Snapshot, snapshotAtEpochNumber uint64) (*GlobalParamsEntry, error) {
+	// Retrieve from db.
+	key := DBKeyForSnapshotGlobalParamsEntry(snapshotAtEpochNumber)
+	globalParamsEntryBytes, err := DBGetWithTxn(txn, snap, key)
+	if err != nil {
+		// We don't want to error if the key isn't found. Instead, return nil.
+		if err == badger.ErrKeyNotFound {
+			return nil, nil
+		}
+		return nil, errors.Wrapf(err, "DBGetSnapshotGlobalParamsEntryWithTxn: problem retrieving value")
+	}
+
+	// Decode from bytes.
+	globalParamsEntry := &GlobalParamsEntry{}
+	rr := bytes.NewReader(globalParamsEntryBytes)
+	if exist, err := DecodeFromBytes(globalParamsEntry, rr); !exist || err != nil {
+		return nil, errors.Wrapf(err, "DBGetSnapshotGlobalParamsEntryWithTxn: problem decoding GlobalParamsEntry: ")
+	}
+	return globalParamsEntry, nil
+}
+
+func DBPutSnapshotGlobalParamsEntryWithTxn(
+	txn *badger.Txn,
+	snap *Snapshot,
+	globalParamsEntry *GlobalParamsEntry,
+	snapshotAtEpochNumber uint64,
+	blockHeight uint64,
+) error {
+	if globalParamsEntry == nil {
+		// This should never happen but is a sanity check.
+		glog.Errorf("DBPutSnapshotGlobalParamsEntryWithTxn: called with nil GlobalParamsEntry, this should never happen")
+		return nil
+	}
+	key := DBKeyForSnapshotGlobalParamsEntry(snapshotAtEpochNumber)
+	return DBSetWithTxn(txn, snap, key, EncodeToBytes(blockHeight, globalParamsEntry))
+}
+
+//
+// SnapshotValidatorEntry
+//
+
+type SnapshotValidatorMapKey struct {
+	SnapshotAtEpochNumber uint64
+	ValidatorPKID         PKID
+}
+
+func (bav *UtxoView) SnapshotCurrentValidators(snapshotAtEpochNumber uint64) error {
```

Comment:
I don't love this but I thought it may be good enough for now. I'm not sure how many validators we are expecting. I load all validators into memory to merge between the database and the UtxoView before snapshotting them. Alternatively, I could copy over all validators in the database one-by-one (without even byte-decoding them), and then just insert or delete any validators in the UtxoView. This would increase the number of db inserts/deletes, but reduce the memory footprint. Happy to make that change if others think it's way better. Also happy to punt on that change until our validator set gets unwieldy to load into memory.